File: ios/source/api/Request.swift
import Foundation

enum Request: Codable, Equatable {
    case get(EntityId)
    case add(EntityId)
    case select(EntityId)
    case remove(EntityId)
    case update(UpdateRequest)
    case transport(TransportRequest)
    case save
    case load(LoadRequest)
    case rename(RenameRequest)
    case beginUpload(BeginUploadRequest)
    case upload(UploadRequest)
    case completeUpload(CompleteUploadRequest)
    case addSample(AddSampleRequest)
    case removeSample(RemoveSampleRequest)
}

struct EntityId: Codable, Equatable {
    var entity: Entity
    var id: Id?
}

enum Entity: String, Codable {
    case all
    case project
    case projects
    case sample
    case section
    case song
    case waveform
}

enum UpdateRequest: Codable, Equatable {
    case song(Song)
    case section(Section)
    case sample(Sample)
}

enum TransportRequest: Codable, Equatable {
    case play
    case stop
    case loop
    case exitLoop
    case queue(QueueRequest)
}

struct LoadRequest: Codable, Equatable {
    var id: Id
}

struct RenameRequest: Codable, Equatable {
    var entity: Entity
    var id: Id?
    var name: String
}

struct BeginUploadRequest: Codable, Equatable {
    var uploadId: Id
    var filename: String
    var format: String
}

struct UploadRequest: Codable, Equatable {
    var uploadId: Id
    var data: Data
}

struct CompleteUploadRequest: Codable, Equatable {
    var uploadId: Id
}

struct AddSampleRequest: Codable, Equatable {
    var songId: Id
    var uploadId: Id
}

struct RemoveSampleRequest: Codable, Equatable {
    var songId: Id
}

struct QueueRequest: Codable, Equatable {
    var songId: Id
    var sectionId: Id
}

File: editor/src/api/request.ts
import {
  AudioFileFormat,
  Entity,
  Project,
  Request,
  Sample,
  Section,
  Song,
  TransportMethod,
} from './bloop';

import {ID, INVALID_ID} from './helpers';

export const getAllRequest = (): Request => {
  return {
    get: {
      entity: Entity.ALL,
      id: INVALID_ID,
    },
  };
};

export const addSectionRequest = (songId: ID): Request => {
  return {
    add: {
      entity: Entity.SECTION,
      id: songId,
    },
  };
};

export const addSectionWithParamsRequest = (
  songId: ID,
  name: string,
  start: number
): Request => {
  return {
    addSection: {
      songId,
      name,
      start,
      loop: false,
      metronome: false,
    },
  };
};

export const addSongRequest = (): Request => {
  return {
    add: {
      entity: Entity.SONG,
      id: INVALID_ID,
    },
  };
};

export const addProjectRequest = (): Request => {
  return {
    add: {
      entity: Entity.PROJECT,
      id: INVALID_ID,
    },
  };
};

export const selectSongRequest = (songId: ID): Request => {
  return {
    select: {
      entity: Entity.SONG,
      id: songId,
    },
  };
};

export const selectSectionRequest = (sectionId: ID): Request => {
  return {
    select: {
      entity: Entity.SECTION,
      id: sectionId,
    },
  };
};

export const removeRequest = (entity: Entity, id: ID): Request => {
  return {
    remove: {
      entity,
      id,
    },
  };
};

export const removeSongRequest = (songId: ID) =>
  removeRequest(Entity.SONG, songId);

export const removeSectionRequest = (sectionId: ID) =>
  removeRequest(Entity.SECTION, sectionId);

export const removeProjectRequest = (projectId: ID) =>
  removeRequest(Entity.PROJECT, projectId);

export const removeSampleRequest = (songId: ID): Request => {
  return {
    removeSample: {
      songId,
    },
  };
};

export const updateSongRequest = (song: Song): Request => {
  return {
    update: {
      song,
    },
  };
};

export const updateSectionRequest = (section: Section): Request => {
  return {
    update: {
      section,
    },
  };
};

export const updateSampleRequest = (sample: Sample): Request => {
  return {
    update: {
      sample,
    },
  };
};

export const updateProjectRequest = (project: Project): Request => {
  return {
    update: {
      project,
    },
  };
};

const transportRequest = (method: TransportMethod): Request => {
  return {
    transport: {
      method,
    },
  };
};

export const playRequest = (): Request =>
  transportRequest(TransportMethod.PLAY);
export const stopRequest = (): Request =>
  transportRequest(TransportMethod.STOP);
export const loopRequest = (): Request =>
  transportRequest(TransportMethod.LOOP);
export const exitLoopRequest = (): Request =>
  transportRequest(TransportMethod.EXIT_LOOP);

export const queueRequest = (songId: ID, sectionId: ID): Request => {
  return {
    transport: {
      method: TransportMethod.QUEUE,
      queue: {songId, sectionId},
    },
  };
};

export const saveRequest = (): Request => {
  return {
    save: {},
  };
};

export const loadProjectsRequest = (): Request => {
  return {
    get: {
      entity: Entity.PROJECTS,
      id: INVALID_ID,
    },
  };
};

export const loadProjectRequest = (projectId: string): Request => {
  return {
    load: {
      projectId,
    },
  };
};

export const renameProjectRequest = (
  projectId: string,
  name: string
): Request => {
  return {
    renameProject: {
      projectId,
      newName: name,
    },
  };
};

export const beginUploadRequest = (
  uploadId: ID,
  filename: string,
  format: AudioFileFormat
): Request => {
  return {
    beginUpload: {
      uploadId,
      filename,
      format,
    },
  };
};

export const uploadRequest = (uploadId: ID, data: ArrayBuffer): Request => {
  return {
    upload: {
      uploadId,
      data: new Uint8Array(data),
    },
  };
};

export const completeUploadRequest = (uploadId: ID): Request => {
  return {
    completeUpload: {
      uploadId,
    },
  };
};

export const addSampleRequest = (songId: ID, uploadId: ID): Request => {
  return {
    addSample: {
      songId,
      uploadId,
    },
  };
};

export const requestWaveformRequest = (sampleId: ID): Request => {
  return {
    get: {
      entity: Entity.SAMPLE,
      id: sampleId,
    },
  };
};
